nexus/src/app/sagas/instance_update/start.rs

```diff
@@ -0,0 +1,214 @@
+            INSTANCE_LOCK_ID,
+            "GenerateInstanceLockId",
+            ACTION_GENERATE_ID.as_ref(),
+        ));
+        builder.append(lock_instance_action());
+        builder.append(fetch_state_and_start_real_saga_action());
+
+        Ok(builder.build()?)
+    }
+}
+
+// start instance update saga: action implementations
+
+async fn siu_lock_instance(
+    sagactx: NexusActionContext,
+) -> Result<Option<instance::UpdaterLock>, ActionError> {
+    let osagactx = sagactx.user_data();
+    let Params { ref serialized_authn, ref authz_instance, .. } =
+        sagactx.saga_params::<Params>()?;
+    let lock_id = sagactx.lookup::<Uuid>(INSTANCE_LOCK_ID)?;
+    let opctx =
+        crate::context::op_context_for_saga_action(&sagactx, serialized_authn);
+    slog::info!(
+        osagactx.log(),
+        "instance update: attempting to lock instance";
+        "instance_id" => %authz_instance.id(),
+        "saga_id" => %lock_id,
+    );
+    let locked = osagactx
+        .datastore()
+        .instance_updater_lock(&opctx, authz_instance, lock_id)
+        .await;
+    match locked {
+        Ok(lock) => Ok(Some(lock)),
+        // Don't return an error if we can't take the lock. This saga will
+        // simply not start the real instance update saga, rather than having to unwind.
+        Err(instance::UpdaterLockError::AlreadyLocked) => Ok(None),
+        // Okay, that's a real error. Time to die!
+        Err(instance::UpdaterLockError::Query(e)) => {
+            Err(ActionError::action_failed(e))
+        }
+    }
+}
+
+async fn siu_lock_instance_undo(
+    sagactx: NexusActionContext,
+) -> Result<(), anyhow::Error> {
+    let Params { ref serialized_authn, ref authz_instance, .. } =
+        sagactx.saga_params::<Params>()?;
+    super::unlock_instance_inner(
+        serialized_authn,
+        authz_instance,
+        &sagactx,
+        None,
+    )
+    .await?;
+    Ok(())
+}
+
+async fn siu_fetch_state_and_start_real_saga(
+    sagactx: NexusActionContext,
+) -> Result<(), ActionError> {
+    let Params { serialized_authn, authz_instance, .. } =
+        sagactx.saga_params::<Params>()?;
+    let osagactx = sagactx.user_data();
+    let lock_id = sagactx.lookup::<Uuid>(INSTANCE_LOCK_ID)?;
+    // Did we get the lock? If so, we can start the next saga, otherwise, just
+    // exit gracefully.
+    let Some(orig_lock) =
+        sagactx.lookup::<Option<instance::UpdaterLock>>(INSTANCE_LOCK)?
+    else {
+        slog::info!(
+            osagactx.log(),
+            "instance update: instance is already locked! doing nothing...";
+            "instance_id" => %authz_instance.id(),
+            "saga_id" => %lock_id,
+        );
+        return Ok(());
+    };
+
+    let opctx =
+        crate::context::op_context_for_saga_action(&sagactx, &serialized_authn);
+
+    let state = osagactx
+        .datastore()
+        .instance_fetch_all(&opctx, &authz_instance)
+        .await
+        .map_err(ActionError::action_failed)?;
```

Comment:
This looks like another place where it'd be nice, if possible, to have the previous update query (i.e. the one that acquired the lock) return the new record instead of having to refetch it.

Reply:
Yeah, for sure --- I kinda punted on that for now to get things working, but I'd like to come back and combine the queries once I've ironed out all the more important stuff.

One thing that I'm not totally sure about is whether it's worth doing all the JOINs necessary to fetch the whole snapshot in a query that might just see "oh it's already locked" and bail. But, OTTOH, we might end up using that data to figure out whether we need to ensure an update saga is scheduled to run once the lock has been released...

Reply:
I don't actually think we can do that, since (IIUC) you can't do joins in a `SELECT FOR UPDATE`[^1], so all the unlock query could return is the instance record itself, and not the VMM and migration records.

[^1]: At least per my googling? You certainly can't represent it in Diesel. If I'm wrong about this, please let me know!

Reply:
@gjcolombo I'm closing this suggestion because, unfortunately, I don't think we can
do this in one query. We need to be able to do LEFT JOINs on the VMM and
migration IDs, and as far as I know, we can't return those rows in an UPDATE
query (please correct me if I'm wrong about SQL here!).

Potentially, we could change the datastore methods to execute the locking (`UPDATE`) query
and the fetch all records (`SELECT ... FROM instance LEFT JOIN vmm LEFT JOIN
migration`) in a single transaction. But, then we're still doing two separate
queries, and I'm a little bit sketched out about mixing transactions with sagas,
especially I'm not sure if the additional complexity gets us anything valuable?
If you disagree, let me know.